src/model_theory/terms_and_formulas.lean

```diff
@@ -362,6 +379,137 @@ begin
       { exact (dif_neg (λ h', h (nat.lt_of_add_lt_add_left h'))).trans (dif_neg h).symm } } }
 end
 
+/-- An atomic formula is either equality or a relation symbol applied to terms. -/
+inductive is_atomic : L.bounded_formula α n → Prop
+| equal {t₁ t₂ : L.term (α ⊕ fin n)} : is_atomic (bd_equal t₁ t₂)
+| rel {l : ℕ} {R : L.relations l} {ts : fin l → L.term (α ⊕ fin n)} :
+    is_atomic (R.bounded_formula ts)
+
+lemma is_atomic.induction {P : L.bounded_formula α n → Prop} {φ : L.bounded_formula α n}
+  (h : is_atomic φ)
+  (he : ∀ (t₁ t₂ : L.term (α ⊕ fin n)), P (bd_equal t₁ t₂))
+  (hr : ∀ {l : ℕ} (R : L.relations l) (ts : fin l → L.term (α ⊕ fin n)), P (R.bounded_formula ts)) :
+  P φ :=
+begin
+  induction h,
+  { apply he },
+  { apply hr }
+end
+
+lemma is_atomic_bd_equal (t₁ t₂ : L.term (α ⊕ fin n)) :
+  is_atomic (t₁.bd_equal t₂ : L.bounded_formula α n) :=
+is_atomic.equal
+
+lemma is_atomic_rel {m} (R : L.relations m) (ts) :
+  is_atomic (R.bounded_formula ts : L.bounded_formula α n) :=
+is_atomic.rel
```

Comment:
Can't you modify the implicitness of the fields of `is_atomic` and then use the fields `is_atomic.equal` (renamed to `is_atomic.bd_equal`?) and `is_atomic.rel` directly? Or is it the naming convention that these should use underscores?